### DISCUSSION TITLE
CRLF in source files - Configuration Standard?

### DISCUSSION THREAD

**Discussion by scaraebeus** [icloud-photos-downloader/icloud_photos_downloader#744] [General]
@AndreyNikiforov - I'm noticing some changes showing up with inconsistent CRLF.  The latest commit in master for example showed the entire `pyicloud/services/photos.py` file being changed, even through there were only two lines.  I can hide this when comparing easily enough with `git diff -w` of course.

I'm using Linux for developing and I imagine others may be using Windows or MacOS in some cases. Should there be a [.gitattributes](https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings#example) to help maintain the correct CRLF settings (to whatever is appropriate) for the repository in case the local `core.autocrlf` settings may vary between contributors?

Is there a recommended CRLF standard we should adhere to, or is it not really that important and can be ignored (using `git diff -w` when needed)?

**Comment by AndreyNikiforov** [comment]
Ugh, my bad, I did not notice CRLF change for the photos.py file. I develop on Windows in linux container with Git client on Windows itself; and on codespaces (==linux container as well). Suspect that my Windows git client added CRLF to photos.py, need to experiment...

So far in icloudpd project I found scripts for unix to be the most sensitive for line-ending, not the python source code. Since you seems familiar with CRLF management in git more than I do, can you pls outlining a) what is the current state of CRLF for the project, b) what kind of issues we are facing or may face, c) what are resolution options with pro/cons, and d) your recommendation?

**Reply by scaraebeus** [reply]
I should note that I'm not an expert in this, nor am I a developer by trade.  :smile:  That being said, I do like to dig into things and figure out the inner workings.

From what I do understand from the research I've done, this is _mostly_ a diff issue, and as you noted, this does affect certain scripts that need to be run on *nix - notably Docker containers with scripts calling `/bin/sh` or `bin/bash` (they'd end up trying to call `bin/sh^M` for example).

a) Current state appears to be a bit of a mix - some files with LF, others with CRLF, others with a mix of both.

b) As python itself doesn't care about what is used (CR/LF/CRLF), and the choice of editor doesn't typically care either, this doesn't really pose too much of an issue at first glance.  However, if not caught, it could create issue as described for any script files that need to be run if not caught.

Additionally, when merging changes, diffs could show a lot of changes that aren't really there, just because the LF changed to CRLF (or vice versa I'd imagine)  - This is more of a nuisance then anything - using `git diff -w` will cause the diff to ignore this when comparing - however, the initial merge will still show a lot of changes as far as I'm aware, making it look like a lot has changed when it hasn't.

c) Options
- Do nothing
  - Pros: No effort
  - Cons: Potential to break scripts if not careful, Large diffs for some contributors even with little changes

- Update the contributing guidelines to specify a standard (for cross-environment development, the standard is to use LF)
  - Require all contributors to ensure they have set core.autocrlf = true if using Windows
  - Pros: Puts the effort on the individual contributor - no changes to the repository
  - Cons: Relies on the individual contributors to do this, may put extra effort on merging pull requests to ensure this was actually done.
 
- Create [.gitattributes](https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings) file and set text files to be converted from CRLF to LF.
  - Pros: Let's Git do all the heavy lifting and ensures a standard is maintained
  - Cons: Requires setting up a .gitattributes file, likely will require a refresh on the repository to get all files aligned to the standard
 
d) My recommendation would be to go with the .gitattributes solution.  It may be a bit of upfront work, but in the long-run will ensure all files are maintained with LF in the repository itself regardless of who is contributing and what platform they are on.

A good write up can be found here for reference . . . https://adaptivepatchwork.com/2012/03/01/mind-the-end-of-your-line/

**Comment by AndreyNikiforov** [comment]
#815
